src/hyperlink/_url.py

```diff
@@ -1705,59 +1822,85 @@ def child(self, *segments):
         new_segs = [_encode_reserved(s) for s in segments]
         return self.__class__(self._url.child(*new_segs))
 
-    def normalize(self, *a, **kw):
+    def normalize(self, scheme=True, host=True, path=True, query=True,
+                  fragment=True, userinfo=True, percents=True):
+        # type: (bool, bool, bool, bool, bool, bool, bool) -> DecodedURL
         """Return a new `DecodedURL` wrapping the result of
         :meth:`~hyperlink.URL.normalize()`
         """
-        return self.__class__(self._url.normalize(*a, **kw))
+        return self.__class__(self._url.normalize(
+            scheme, host, path, query, fragment, userinfo, percents
+        ))
 
     @property
     def absolute(self):
+        # type: () -> bool
         return self._url.absolute
 
     @property
     def scheme(self):
+        # type: () -> Text
         return self._url.scheme
 
     @property
     def host(self):
+        # type: () -> Text
         return _decode_host(self._url.host)
 
     @property
     def port(self):
+        # type: () -> Optional[int]
         return self._url.port
 
     @property
     def rooted(self):
+        # type: () -> bool
         return self._url.rooted
 
     @property
     def path(self):
+        # type: () -> Sequence[Text]
         try:
-            return self._path
+            return cast(
+                Tuple[Text, ...],
+                self._path  # type: ignore[has-type] can't determine
+            )
         except AttributeError:
             pass
-        self._path = tuple([_percent_decode(p, raise_subencoding_exc=True)
-                            for p in self._url.path])
+        self._path = tuple([
+            _percent_decode(p, raise_subencoding_exc=True)
+            for p in self._url.path
+        ])
         return self._path
 
     @property
     def query(self):
+        # type: () -> QueryPairs
         try:
-            return self._query
+            return cast(
+                QueryPairs,
```

Comment:
Dang, are these casts in the critical path logic necessary? On a previous readthrough I had only seen the ones in the test file. 

I remember some past PRs getting a slight grilling for the potential perf impact of converting between generator expressions and list comprehensions, is there somewhere I can learn about the tradeoffs of using this approach, or do you mind giving me a quick primer? Like, is mypy not able to infer these as `(str, obj)` or `(str, str/None)`?

Reply:
Like, pretty sure pytype can do it, especially considering all the other annotations you've put in: https://github.com/google/pytype

Are we wedded to mypy as the checker, or would you be open to trying pytype (I've heard good things, admittedly, from googlers, so I'm cautious too)

Reply:
`cast()` is supposed to be equivalent to `lambda x: x` at runtime, so there's a function call of overhead, and that's is.  But that's not zero.

I'm pretty annoyed that some of these casts are necessary to appease `mypy`… I think one option is that we can do more `# type: ignore` hooey here and wait on `mypy` to fix what I suspect are  some bugs instead of casting here.  This was an attempt to narrow down the ignores to a minimum, then see if `mypy` can incrementally address the issues.

Reply:
OK I got rid of a bunch of calls to `cast()`